README.md
# Elara

A command line utility for processing (in big batches or bit by bit) MATSim XML outputs (events or plans files) into useful tabular and geospatial outputs for analysis.

## Contents

* [Introduction](https://github.com/arup-group/elara#introduction)
* [Installation](https://github.com/arup-group/elara#installation)
* [Inputs](https://github.com/arup-group/elara#inputs)
* [Output Handlers](https://github.com/arup-group/elara#output-handlers)
* [Running from a Configuration File](https://github.com/arup-group/elara#running-from-a-configuration-file)
* [Command Line Reference](https://github.com/arup-group/elara#command-line-reference)
* [Configuration Options](https://github.com/arup-group/elara#configuration-options)
* [Tests](https://github.com/arup-group/elara#tests)
* [Debug](https://github.com/arup-group/elara#debug)
* [About](https://github.com/arup-group/elara#about)
* [Adding Features](https://github.com/arup-group/elara#adding-features)
* [Todo](https://github.com/arup-group/elara#todo)
* [What does the name mean?](https://github.com/arup-group/elara#what-does-the-name-mean)

## Introduction

Elara allows processing of complex MATSim outputs into more easilly useable forms and formats. For example extracting hourly flows of vehicles for all links in a network (`link_vehicle_counts`). Elara outputs are typically some form of aggregation of simulation outputs. Elara outputs are typically made available in tabular (`csv`) and spatial (`geojson`) formats. Spatial representations are converted to EPSG:4326, which works in kepler.

Once installed Elara is intended to be used as a Command Line Interface (CLI), within this interface you can choose to run elara via a [config](https://github.com/arup-group/elara#configuration) or purely through the [CLI](https://github.com/arup-group/elara#command-line-reference) using command line arguments. The CLI is preferred for producing single outputs, for example to extract vehicle kms, whereas using
a config can handle big batches of outputs in one go. So if you are after a quick output - best to use the CLI, whereas if you need a collection of outputs and/or reproducibility, a config is preferred.

The CLI and configs share naming conventions and structure. But we recommend learning about configs first as this provides an overview of most available tools.

## Installation

Assuming Python >=3.7, virtualenv and using git, clone or download this repository.

```{sh}
git clone git@github.com:arup-group/elara.git
```

Once available locally, navigate to the folder to install Elara and its dependencies. Using a virtual environment is highly recommended.

### OSX

```{sh}
cd elara
virtualenv -p python3.7 venv
source venv/bin/activate
pip3 install -e .
pytest
elara --help
```

### Windows

It is recommended to use an Anaconda environment for installation on Windows:

```{sh}
conda create -n elara python=3.7
conda activate elara
conda install geopandas
git clone git@github.com:arup-group/elara.git
cd elara
pip install -e .
pytest
elara --help
```

On Windows, installation may fail due to failure to locate underlying geospatial dependencies (with a traceback indicating `geos_c.dll` cannot be found, or similar). In this case, it is recommended to first install `geopandas` from conda-forge.

```{sh}
conda install geopandas -c conda-forge
```

After removing `geopandas`, `fiona`, and `shapely` from the `requirements.txt` file, install Elara as per above.

## Inputs

Inputs to Elara are MATSim format output files, eg:

* **events** = "./tests/test_fixtures/output_events.xml"
* **network** = "./tests/test_fixtures/output_network.xml"
* **transit_schedule** = "./tests/test_fixtures/output_transitSchedule.xml"
* **transit_vehicles** = "./tests/test_fixtures/output_transitVehicles.xml"
* **attributes** = "./tests/test_fixtures/output_personAttributes.xml"
* **plans** = "./tests/test_fixtures/output_plans.xml"
* **output_config_path** = "./tests/test_fixtures/output_config.xml"
* **road pricing config** = "./tests/test_fixtures/roadpricing.xml"
* **vehicles** = "./tests/test_fixtures/output_vehicles.xml"

Depending on what outputs you require from elara, some of these inputs may not be required, but it is often convenient to have them all available as a default. In most cases these Elara inputs may be compressed using gzip (`file.xml.gz`).

The road pricing config is required when using the (deprecated) ``toll_log`` plan handler, but *not* when calcuating toll logs using the `agent_tolls_log` event handler.

## Output Handlers

Elara supports and can be selectively configured to output a growing number of outputs. The units responsible
for each output are often referred to as 'handlers', although `elara.factory` describes them more generally as `Tools`.

There are four main types of handler/tools, arranged into python modules:

* elara.event_handlers
* elara.plan_handlers
* elara.postprocessing
* elara.benchmarking

Within each module a `WorkStation` is responsible for orchestrating all that modules tools. The `elara.factory` module is then responsible for orchestratinig all the `Workstations` together.

### **Event Handlers/WorkStation Tools**

*Users are encouraged to add new tools as they require. The below lists are updated periodically but may not be complete.*

These are processed by streaming (in order) through all output events from simulation.

Currently supported handlers include:

* ``link_vehicle_counts``: Produce link volume counts and volume capacity ratios by time slice. Counts **vehicles** entering link (PT vehicles counted once).
* ``link_vehicle_capacity``: Produce link capacity counts by time slice for understanding PT crowding. Sums the vehicle capacities of all **vehicles** entering link for the particular PT mode.
* ``link_passenger_counts``: Produce link passenger counts by time slice. Counts **agents** entering link.
* ``link_vehicle_speeds``: Produce average vehicle speeds across link (in kilometers per hour).
* ``route_passenger_counts``: (WIP) Produce vehicle occupancies by transit routes.
* ``stop_passenger_interactions``: Boardings and Alightings by time slice.
* ``stop_to_stop_passenger_counts``: Passenger counts between directly connected stops/stations.
* ``stop_passenger_waiting``: Agent waiting times for unique pt interaction events.
* ``vehicle_departure_log``: Vehicle departures and delays from facilities (stops in the case of PT).
* ``vehicle_passenger_log``: A log of every passenger boarding and alighting to/from a transit vehicle.
* ``vehicle_passenger_graph``: Experimental support for building interaction graph objects (networkx).
* ``agent_tolls_log``: Produce a log and 24-hour summary (total amount, number of tolls incurred) of agents' tolling events.

### **Plan Handlers/WorkStation Tools**

These are processed by streaming through all output plans from simulation.
Compared to the event based outputs these are typically more aggregate but can be computationally faster and can be used to expose agent plan
'memories' and plan scoring.

Currently supported handlers include:

* ``trip_modes``: Produce modeshares and counts of all trips.
* ``trip_activity_modes``: Produce modeshares and counts of all trips to specified destination activities.
* ``plan_modes``: Produce modeshares and counts of all plans (ie the dominant mode for each person).
* ``plan_activity_modes``: Produce modeshares and counts of all plans (ie the dominant mode for each person) to specified destination activities.
* ``trip_logs``: Produce agent activity logs and trip logs for all selected plans. Omits pt interactions and individual trip legs. Trip mode is based on maximum leg distance.
* ``leg_logs``: Produce agent activity logs and leg logs for all selected plans.
* ``plan_logs``: Produce agent plans including unselected plans and scores.
* ``agent_highway_distance_logs``: Produce agent distances by car on different road types. Requires network to have `osm:way:highways` attribute.
* ``trip_highway_distance_logs``: Produce flat output of agent trip distances by car on different road types. Requires network to have `osm:way:highways` attribute.
* **(DEPRECATED)** ``toll_logs``: Produces summary of amounts agents paid at tolls based on route information contained in agent plans. Requires road pricing input file. Only works for option ``["car"]``. *The AgentTollsPaidFromRPConfig handler is still supported, but has been superseded by an event handler. This handler calculates toll payments via a link-lookup with the road pricing configuration file, and will not account for any in-simulation adjustments for differential or capped road pricing.*

### **Post Processing Handlers/Workstation Tools**

These are outputs produced through additional post-processing of the above outputs. Currently supported postprocessors include:

* ``vkt``: Produce link volume vehicle kms by time slice.
* ``trip_duration_breakdown``: Produce binned trip durations.
* ``trip_euclid_distance_breakdown``: Produce binned trip distances.
* ``plan_summary``: Produce summary statistics for all agent plans

### **Benchmarking Handlers/WorkStation Tools**

Elara can also assist with validation or 'benchmarking' of simulations. Elara will compare and present simulation results from the above outputs to available benchmarks, it will aditionally output a distance based score for the model. Where distance is some measure of how different the simulation is from the observed data.

The handlers in this module require correctly formatted benchmark data. Examples can be found in the project `example_benchmark_data` folder. This documentation has been moved to [here](https://github.com/arup-group/elara/benchmarks.md).

## Running from a Configuration File

For reproducibility or to process a range of outputs, using a config file is the most sensible and processing efficient way to use Elara. From your terminal:

`elara run <CONFIG PATH>`

Config files must be `.toml` format and be roughly formatted as follows. The various fields are detailed further below and an [example config](https://github.com/arup-group/elara/blob/master/example_configs/config.toml) is also included in the repo:

```{.toml}
[scenario]
name = "test_town"
time_periods = 24
scale_factor = 0.01
crs = "EPSG:27700"

[inputs]
events = "./tests/test_fixtures/output_events.xml"
network = "./tests/test_fixtures/output_network.xml"
transit_schedule = "./tests/test_fixtures/output_transitSchedule.xml"
transit_vehicles = "./tests/test_fixtures/output_transitVehicles.xml"
attributes = "./tests/test_fixtures/output_personAttributes.xml"
plans= "./tests/test_fixtures/output_plans.xml"
output_config_path = "./tests/test_fixtures/output_config.xml"
road_pricing = "./tests/test_fixtures/roadpricing.xml"
vehicles = "./tests/test_fixtures/output_vehicles.xml"

[outputs]
path = "./tests/test_outputs"

[event_handlers]
link_vehicle_counts = ["car", "bus"]
link_passenger_counts = ["bus"]
stop_passenger_counts = ["bus"]
stop_passenger_waiting = ["all"]

[plan_handlers]
trip_modes = {modes=["all"]}
trip_activity_mode_shares = {destination_activity_filters = ["work"]}
trip_logs = ["all"]
agent_highway_distance_logs = ["car"]
trip_highway_distance_logs = ["car"]

[post_processors]
vkt = ["car"]
plan_summary = ["all"]
trip_duration_breakdown = ["all"]
trip_euclid_distance_breakdown = ["all"]
```

You can run this config on some toy data: `elara run example_configs/config.toml` (from the project root).

If your MATSim outputs use default names and are in the same directory, you may optionally pass the path of this directory as single argument to `[inputs]` using, e.g. `inputs_directory = "./tests/test_fixtures/"`. **NB:** If using the ``toll_log`` plan handler, you must still provide the `road_pricing` path separately.

### Configuration dry run

`elara run` supports some optional flags and arguments which can be discovered with `elara run --help`, most useful is `elara run -d` (or `elara run --dry`) which can be use to test a config without providing valid inputs or undertaking the processing.

## Command Line Reference

Elara can also be more generally used via the CLI to process individual outputs. Used in this manner the CLI should be pretty discoverable, once installed, try the command `elara` in your terminal to find out about the available options:

```{sh}
$ elara
Usage: elara [OPTIONS] COMMAND [ARGS]...

  Command line tool for processing a MATSim scenario events output.

Options:
  --help  Show this message and exit.

Commands:
  benchmarks       Access benchmarks output group.
  event-handlers   Access event handler output group.
  plan-handlers    Access plan handler output group.
  post-processors  Access post processing output group.
  run              Run Elara using a config.toml, examples are included in...

```

Further commands can then be explored. In general the commands and options available follow the same structure as configuration.

```{sh}
$ elara event-handlers volume-counts --help

Usage: elara event-handlers volume-counts [OPTIONS] MODES...

  Create a volume counts output for a given mode or modes. Example
  invocation for "car" and "bus" modes with name "test" and scale factor at
  20% is:

  $ elara event-handlers volume-counts car bus -n test -s .2

Options:

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.0]

## [0.1.0] - 2023-06-23

### Added

- Verbose option changed to 'debug' (#150), this is not a breaking change but old config files using `verbose = true` or some equivalent will now silently default to `debug = false` which is equivalent to the old `verbose = false`.


elara/main.py
import os.path
import click
import logging
from pathlib import Path
from elara.helpers import PathPath, NaturalOrderGroup
from elara.config import Config, RequirementsWorkStation, PathFinderWorkStation
from elara.inputs import InputsWorkStation
from elara.plan_handlers import PlanHandlerWorkStation
from elara.input_plan_handlers import InputPlanHandlerWorkstation
from elara.event_handlers import EventHandlerWorkStation
from elara.postprocessing import PostProcessWorkStation
from elara.benchmarking import BenchmarkWorkStation
from elara import factory


@click.group(cls=NaturalOrderGroup)
def cli():
    """
    Command line tool for processing a MATSim scenario events output.
    """
    pass


@cli.command()
@click.argument("config_path", type=click.Path(exists=True))
@click.option("--dry", "-d", is_flag=True, help="test run a config file.")
@click.option("--path_override", '-o', default=None, help="over-ride input path root.")
@click.option("--root", '-r', default=None, help="over-ride all path roots.")
@click.option("--output_directory_override", default=None, help="over-ride output directory to new path.")
def run(config_path, dry, path_override, root, output_directory_override):
    """
    Run Elara using a config.
    :param config_path: Configuration file path
    :param path_override: containing directory to update for [inputs], outputs.path in toml
    :param root: add root to all paths (assumes that paths in config are relative)
    :param output_directory_override: change outputs directory
    :param dry: flag to initiate a run test
    """

    if path_override and root:
        raise UserWarning(
            "Cannot run elara from config with both --path_override and --root options, please choose one."
            )

    config = Config(config_path)

    if path_override:
        config.override(path_override)

    if root:
        config.set_paths_root(root)

    if output_directory_override:
        config.output_directory_override(output_directory_override)

    """
    Main logic:
        1) define workstations
        2) connect workstations (define dependencies)
        3) build all resulting graph requirements
    :param config: Session configuration object
    """

    main(config=config, dry_run=dry)


def main(config, dry_run=False) -> None:
    logger = logging.getLogger(__name__)
    logger.setLevel(config.logging)

    handler = logging.StreamHandler()
    formatter = logging.Formatter('%(asctime)s %(name)-12s %(levelname)-3s %(message)s', '%m-%d %H:%M')
    handler.setFormatter(formatter)

    logger.info('Starting')
    requirements = define_and_connect_workstations(config, logger)
    if dry_run:
        factory.dry_run_build(requirements)
    else:
        factory.build(requirements)
    logger.info('Done')


def define_and_connect_workstations(config, logger):

    config.experienced_plans_warning()

    # Create output folder if it does not exist
    if not os.path.exists(config.output_path):
        logger.info(f'Creating new output directory {config.output_path}')
        os.makedirs(config.output_path)

    # 1: Define Work Stations
    config_requirements = RequirementsWorkStation(config)
    postprocessing = PostProcessWorkStation(config)
    benchmarks = BenchmarkWorkStation(config)
    event_handlers = EventHandlerWorkStation(config)
    input_plan_handlers = InputPlanHandlerWorkstation(config)
    plan_handlers = PlanHandlerWorkStation(config)
    input_workstation = InputsWorkStation(config)
    paths = PathFinderWorkStation(config)

    # 2: Connect Workstations
    config_requirements.connect(
        managers=None,
        suppliers=[postprocessing, benchmarks, event_handlers, plan_handlers, input_plan_handlers]
    )
    benchmarks.connect(
        managers=[config_requirements],
        suppliers=[postprocessing, event_handlers, plan_handlers, input_plan_handlers],
    )
    postprocessing.connect(
        managers=[config_requirements, benchmarks],
        suppliers=[input_workstation, event_handlers, plan_handlers, input_plan_handlers]
    )
    event_handlers.connect(
        managers=[postprocessing, benchmarks, config_requirements],
        suppliers=[input_workstation]
    )
    plan_handlers.connect(
        managers=[config_requirements, benchmarks, postprocessing],
        suppliers=[input_workstation]
    )
    input_plan_handlers.connect(
        managers=[config_requirements, benchmarks, postprocessing],
        suppliers=[input_workstation]
    )
    input_workstation.connect(
        managers=[event_handlers, plan_handlers, input_plan_handlers, postprocessing],
        suppliers=[paths]
    )
    paths.connect(
        managers=[input_workstation],
        suppliers=None
    )
    return config_requirements


def common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    using_experienced_plans = False if no_experienced_plans else True
    plans_file = "output_experienced_plans.xml.gz" if using_experienced_plans else "output_plans.xml.gz"
    attributes_file = "output_plans.xml.gz" if version >= 12 else "output_personAttributes.xml.gz"

    return {
        "scenario":
            {
                "name": name,
                "time_periods": time_periods,
                "scale_factor": scale_factor,
                "version": version,
                "crs": epsg,
                "debug": debug,
                "using_experienced_plans": using_experienced_plans
            },
        "inputs":
            {
                "events": inputs_path / "output_events.xml.gz",
                "network": inputs_path / "output_network.xml.gz",
                "transit_schedule": inputs_path / "output_transitSchedule.xml.gz",
                "transit_vehicles": inputs_path / "output_transitVehicles.xml.gz",
                "attributes": inputs_path / attributes_file,
                "plans": inputs_path / plans_file,
                "output_config_path": inputs_path / "output_config.xml",
            },
        "event_handlers":
            {
             },
        "plan_handlers":
            {
            },
        "post_processors":
            {
            },
        "benchmarks":
            {
            },
        "outputs":
            {
                "path": outputs_path,
                "contract": not full,
            },
    }


def common_options(func):
    func = click.option(
        '-d', '--debug', is_flag=True, help="Switch on debug verbosity."
    )(func)

    func = click.option(
        '-n', '--name', type=click.STRING, default=str(Path(os.getcwd())).split('/')[-1],
        help="Scenario name, defaults to root dir name."
    )(func)

    func = click.option(
        '-x', '--no_experienced_plans', is_flag=True, default=False,
        help="Switch for turning off Experienced Plans. Set to use output_plans instead."
    )(func)

    func = click.option(
        '-i', '--inputs_path', type=PathPath(exists=True), default=Path(os.getcwd()),
        help="Inputs path location, defaults to current root."
    )(func)

    func = click.option(
        '-o', '--outputs_path', type=PathPath(), default=Path(os.getcwd()) / 'elara_out',
        help="Outputs path, defaults to './elara_out'."
    )(func)

    func = click.option(
        '-p', '--time_periods', type=click.INT, default=24,
        help="Time period breakdown, defaults to 24 (hourly."
    )(func)

    func = click.option(
        '-s', '--scale_factor', type=click.FLOAT, default=0.1,
        help="Scale factor, defaults to 0.1 (10%)."
    )(func)

    func = click.option(
        '-v', '--version', type=click.INT, default=12,
        help="MATSim version {11,12}"
    )(func)

    func = click.option(
        '-e', '--epsg', type=click.STRING, default="EPSG:27700",
        help="EPSG string, defaults to 'EPSG:27700' (UK)."
    )(func)

    func = click.option(
        '-f', '--full', is_flag=True, default=True,
        help="Option to disable output contracting."
    )(func)

    return func


@cli.group(name="event-handlers", cls=NaturalOrderGroup)
def event_handlers():
    """
    Access event handler output group.
    """
    pass


@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def link_vehicle_counts(
        modes, debug, no_experienced_plans, name, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a volume counts output for a given mode or modes. Example invocation for "car" and "bus"
    modes with name "test" and scale factor at 20% is:

    $ elara event-handlers link-vehicle-counts car bus -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["link_vehicle_counts"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)

@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def link_vehicle_capacity(
        modes, debug, no_experienced_plans, name, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a link capacity output for a given mode or modes. Example invocation for "car" and "bus"
    modes with name "test" and scale factor at 20% is:

    $ elara event-handlers link-vehicle-counts car bus -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["link_vehicle_capacity"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)

@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def link_passenger_counts(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a passenger counts output for a given mode or modes. Example invocation for "train" and
    "bus" modes with name "test" and scale factor at 20% is:

    $ elara event-handlers link-passenger-counts train bus -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["link_passenger_counts"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def route_passenger_counts(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a route passenger counts output for a given mode or modes. Example invocation for "train" and
    "bus" modes with name "test" and scale factor at 20% is:

    $ elara event-handlers route-passenger-counts train bus -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans,inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["route_passenger_counts"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def stop_passenger_counts(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a stop interactions output for a given mode or modes. Example invocation for "train" and
    "bus" modes with name "test" and scale factor at 20% is:

    $ elara event-handlers stop-passenger-counts train bus -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["stop_passenger_counts"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def stop_to_stop_passenger_counts(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create stop to stop passenger counts output for a given mode or modes. Example invocation for "train" and
    "bus" modes with name "test" and scale factor at 20% is:

    $ elara event-handlers stop_to_stop_passenger_counts train bus -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["stop_to_stop_passenger_counts"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def vehicle_link_log(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    create a vehicle link log

    $ elara event-handlers vehicle_link_log train bus -n test
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["vehicle_link_log"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@event_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def vehicle_links_animate(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    create a vehicle animation arrow format file

    $ elara event-handlers vehicle_links_animate all -n test
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["event_handlers"]["vehicle_links_animate"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@cli.group(name="plan-handlers", cls=NaturalOrderGroup)
def plan_handlers():
    """
    Access plan handler output group.
    """
    pass


@plan_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def mode_shares(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a mode share output for a given option. Example invocation for option "all" and
     scale factor at 20% is:

    $ elara plan-handlers mode-shares all -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["mode_shares"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)

@plan_handlers.command()
@click.argument('destination_activity_filters', nargs=-1, type=click.STRING, required=True)
@common_options
def trip_destination_mode_share(
        debug, name, no_experienced_plans, inputs_path, outputs_path, destination_activity_filters, time_periods, scale_factor, version, epsg, full
):
    """
    Create a mode share output for a given option. Example invocation for option "work" and
     scale factor at 20% is:

    $ elara plan-handlers activity-mode-shares work -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["trip_destination_mode_share"] = {'modes': ['all'], 'destination_activity_filters':destination_activity_filters}
    config = Config(override=override)
    main(config)

@plan_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def leg_logs(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create an agent leg logs output. Example invocation for option "all" is:

    $ elara plan-handlers leg-logs all
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["leg_logs"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@plan_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def trip_logs(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create an agent trip logs output. Example invocation for option "all" is:

    $ elara plan-handlers trip-logs all
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["trip_logs"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@plan_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def utility_logs(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create an agent plan utility output. Example invocation for option "all" is:

    $ elara plan-handlers utility_logs all
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["utility_logs"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@plan_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def plan_logs(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create an agent plans output for a given option. Example invocation for sub-populations "a",
    "b" and "c" with name "test" and scale factor at 20% is:

    $ elara plan-handlers plan_logs a b c -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["plan_logs"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@plan_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def agent_highway_distance_logs(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create an agent highways distances output for a given option. Example invocation for option "car" name
    "test" and scale factor at 20% is:

    $ elara plan-handlers agent_highway_distance_logs car -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["agent_highway_distance_logs"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@plan_handlers.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def trip_highway_distance_logs(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a trip highways distances output for a given option. Example invocation for option "car" name
    "test" and scale factor at 20% is:

    $ elara plan-handlers trip_highway_distance_logs car -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["plan_handlers"]["trip_highway_distance_logs"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)


@cli.group(name="post-processors", cls=NaturalOrderGroup)
def post_processors():
    """
    Access post processing output group.
    """
    pass


@post_processors.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def vkt(
        modes, debug, no_experienced_plans, name, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a VKT output for a given mode or modes. Example invocation for mode "car", name
    "test" and scale factor at 20% is:

    $ elara plan-processors vkt car -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["post_processors"]["vkt"] = list(modes)
    config = Config(override=override)
    main(config)


@post_processors.command()
@click.argument('modes', nargs=-1, type=click.STRING, required=True)
@common_options
def plan_summary(
        modes, debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
):
    """
    Create a Plan Summary output. Example invocation for option "all", name
    "test" and scale factor at 20% is:

    $ elara plan-processors plan-summary all -n test -s .2
    """
    override = common_override(
        debug, name, no_experienced_plans, inputs_path, outputs_path, time_periods, scale_factor, version, epsg, full
    )
    override["post_processors"]["plan_summary"] = {'modes': list(modes)}
    config = Config(override=override)
    main(config)
